server/Mac/mf_audin.c

```diff
@@ -29,47 +29,139 @@
 #include <freerdp/log.h>
 #define TAG SERVER_TAG("mac")
 
-/**
- * Function description
- *
- * @return 0 on success, otherwise a Win32 error code
- */
-static UINT mf_peer_audin_opening(audin_server_context* context)
+static UINT mf_peer_audin_receive_version(audin_server_context* audin, const SNDIN_VERSION* version)
+{
+	mfPeerContext* context = audin->userdata;
+	SNDIN_FORMATS formats = { 0 };
+
+	if (version->Version == 0)
+	{
+		WLog_ERR(TAG, "Received invalid AUDIO_INPUT version from client");
+		return ERROR_INVALID_DATA;
+	}
+
+	WLog_DBG(TAG, "AUDIO_INPUT version of client: %u", version->Version);
+
+	formats.NumFormats = context->audin_n_server_formats;
+	formats.SoundFormats = context->audin_server_formats;
+
+	return audin->SendFormats(audin, &formats);
+}
+
+static UINT send_open(audin_server_context* audin)
+{
+	mfPeerContext* context = audin->userdata;
+	SNDIN_OPEN open = { 0 };
+
+	open.FramesPerPacket = 441;
+	open.initialFormat = context->audin_client_format_idx;
+	open.captureFormat.wFormatTag = WAVE_FORMAT_PCM;
+	open.captureFormat.nChannels = 2;
+	open.captureFormat.nSamplesPerSec = 44100;
+	open.captureFormat.nAvgBytesPerSec = 44100 * 2 * 2;
+	open.captureFormat.nBlockAlign = 4;
+	open.captureFormat.wBitsPerSample = 16;
+
+	return audin->SendOpen(audin, &open);
+}
+
+static UINT mf_peer_audin_receive_formats(audin_server_context* audin, const SNDIN_FORMATS* formats)
+{
+	mfPeerContext* context = audin->userdata;
```

Comment:
better `WINPR_ASSERT` parameters

Reply:
Done (same with the other `WINPR_ASSERT` parts).